src/System.Private.Uri/src/System/UriExt.cs

```diff
@@ -499,29 +499,46 @@ internal unsafe bool InternalIsWellFormedOriginalString()
         public static string UnescapeDataString(string stringToUnescape)
         {
             if ((object)stringToUnescape == null)
+            {
                 throw new ArgumentNullException(nameof(stringToUnescape));
+            }
 
             if (stringToUnescape.Length == 0)
+            {
                 return string.Empty;
+            }
 
             unsafe
             {
                 fixed (char* pStr = stringToUnescape)
                 {
                     int position;
                     for (position = 0; position < stringToUnescape.Length; ++position)
+                    {
                         if (pStr[position] == '%')
+                        {
                             break;
+                        }
+                    }
 
                     if (position == stringToUnescape.Length)
+                    {
                         return stringToUnescape;
+                    }
 
                     UnescapeMode unescapeMode = UnescapeMode.Unescape | UnescapeMode.UnescapeAll;
                     position = 0;
-                    char[] dest = new char[stringToUnescape.Length];
-                    dest = UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, dest, ref position,
-                        c_DummyChar, c_DummyChar, c_DummyChar, unescapeMode, null, false);
-                    return new string(dest, 0, position);
+                    PooledCharArray pooledArray = new PooledCharArray(stringToUnescape.Length);
+                    UriHelper.UnescapeString(stringToUnescape, 0, stringToUnescape.Length, ref pooledArray, ref position,
+                    c_DummyChar, c_DummyChar, c_DummyChar, unescapeMode, null, false);
+
+                    if(pooledArray.IsSameString(pStr, 0, position))
+                    {
+                        pooledArray.Release();
+                        return stringToUnescape;
+                    }
```

Comment:
as noted in PooledCharArray, I think we can cut this block and just make GetSTringAndRelease do al the work.